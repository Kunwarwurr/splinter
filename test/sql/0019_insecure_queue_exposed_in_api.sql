begin;

  set local search_path = '';
  set local pgrst.db_schemas = 'public';

  -- 0 issues
  select * from lint."0019_insecure_queue_exposed_in_api";

  -- Satisfy all failure conditions
  create schema pgmq;
  create schema pgmq_public;
  create table pgmq.q_foo(id int);
  create table pgmq.q_bar(id int);
  set local pgrst.db_schemas = 'public,pgmq_public';

  grant usage on schema pgmq to anon, authenticated;
  grant usage on schema pgmq_public to anon, authenticated;
  grant select on pgmq.q_foo to anon, authenticated;
  grant select on pgmq.q_bar to anon, authenticated;


  -- 2 issue - because not in schema
  select * from lint."0019_insecure_queue_exposed_in_api";

  savepoint a;

  -- dropping select permission resolves issue
  revoke select on pgmq.q_foo from anon, authenticated;
  revoke select on pgmq.q_bar from anon, authenticated;

  -- 0 issue - because not in schema
  select * from lint."0019_insecure_queue_exposed_in_api";

  -- restore to failure mode
  rollback to savepoint a;
  select count(*) from lint."0019_insecure_queue_exposed_in_api";

  set local pgrst.db_schemas = 'public,pgmq_public';

  -- 0 issue - because not in schema
  select * from lint."0019_insecure_queue_exposed_in_api";

  -- restore to failure mode
  rollback to savepoint a;
  select count(*) from lint."0019_insecure_queue_exposed_in_api";

  alter table pgmq.q_foo enable row level security;
  alter table pgmq.q_bar enable row level security;

  -- 0 issue - because RLS is active
  select * from lint."0019_insecure_queue_exposed_in_api";


rollback;
