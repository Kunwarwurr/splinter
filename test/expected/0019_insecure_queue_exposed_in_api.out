begin;
  set local search_path = '';
  set local pgrst.db_schemas = 'public';
  -- 0 issues
  select * from lint."0019_insecure_queue_exposed_in_api";
 name | title | level | facing | categories | description | detail | remediation | metadata | cache_key 
------+-------+-------+--------+------------+-------------+--------+-------------+----------+-----------
(0 rows)

  -- Satisfy all failure conditions
  create schema pgmq;
  create schema pgmq_public;
  create table pgmq.q_foo(id int);
  create table pgmq.q_bar(id int);
  set local pgrst.db_schemas = 'public,pgmq_public';
  grant usage on schema pgmq to anon, authenticated;
  grant usage on schema pgmq_public to anon, authenticated;
  grant select on pgmq.q_foo to anon, authenticated;
  grant select on pgmq.q_bar to anon, authenticated;
  -- 2 issue - because not in schema
  select * from lint."0019_insecure_queue_exposed_in_api";
             name              |             title             | level |  facing  | categories |                           description                           |                            detail                             |                                            remediation                                            |                       metadata                       |             cache_key             
-------------------------------+-------------------------------+-------+----------+------------+-----------------------------------------------------------------+---------------------------------------------------------------+---------------------------------------------------------------------------------------------------+------------------------------------------------------+-----------------------------------
 insecure_queue_exposed_in_api | Insecure Queue Exposed in API | ERROR | EXTERNAL | {SECURITY} | Detects cases where an insecure Queue is exposed over Data APIs | Table \`pgmq.q_bar\` is public, but RLS has not been enabled. | https://supabase.com/docs/guides/database/database-linter?lint=0019_insecure_queue_exposed_in_api | {"name": "q_bar", "type": "table", "schema": "pgmq"} | rls_disabled_in_public_pgmq_q_bar
 insecure_queue_exposed_in_api | Insecure Queue Exposed in API | ERROR | EXTERNAL | {SECURITY} | Detects cases where an insecure Queue is exposed over Data APIs | Table \`pgmq.q_foo\` is public, but RLS has not been enabled. | https://supabase.com/docs/guides/database/database-linter?lint=0019_insecure_queue_exposed_in_api | {"name": "q_foo", "type": "table", "schema": "pgmq"} | rls_disabled_in_public_pgmq_q_foo
(2 rows)

  savepoint a;
  -- dropping select permission resolves issue
  revoke select on pgmq.q_foo from anon, authenticated;
  revoke select on pgmq.q_bar from anon, authenticated;
  -- 0 issue - because not in schema
  select * from lint."0019_insecure_queue_exposed_in_api";
 name | title | level | facing | categories | description | detail | remediation | metadata | cache_key 
------+-------+-------+--------+------------+-------------+--------+-------------+----------+-----------
(0 rows)

  -- restore to failure mode
  rollback to savepoint a;
  select count(*) from lint."0019_insecure_queue_exposed_in_api";
 count 
-------
     2
(1 row)

  set local pgrst.db_schemas = 'public,pgmq_public';
  -- 0 issue - because not in schema
  select * from lint."0019_insecure_queue_exposed_in_api";
             name              |             title             | level |  facing  | categories |                           description                           |                            detail                             |                                            remediation                                            |                       metadata                       |             cache_key             
-------------------------------+-------------------------------+-------+----------+------------+-----------------------------------------------------------------+---------------------------------------------------------------+---------------------------------------------------------------------------------------------------+------------------------------------------------------+-----------------------------------
 insecure_queue_exposed_in_api | Insecure Queue Exposed in API | ERROR | EXTERNAL | {SECURITY} | Detects cases where an insecure Queue is exposed over Data APIs | Table \`pgmq.q_bar\` is public, but RLS has not been enabled. | https://supabase.com/docs/guides/database/database-linter?lint=0019_insecure_queue_exposed_in_api | {"name": "q_bar", "type": "table", "schema": "pgmq"} | rls_disabled_in_public_pgmq_q_bar
 insecure_queue_exposed_in_api | Insecure Queue Exposed in API | ERROR | EXTERNAL | {SECURITY} | Detects cases where an insecure Queue is exposed over Data APIs | Table \`pgmq.q_foo\` is public, but RLS has not been enabled. | https://supabase.com/docs/guides/database/database-linter?lint=0019_insecure_queue_exposed_in_api | {"name": "q_foo", "type": "table", "schema": "pgmq"} | rls_disabled_in_public_pgmq_q_foo
(2 rows)

  -- restore to failure mode
  rollback to savepoint a;
  select count(*) from lint."0019_insecure_queue_exposed_in_api";
 count 
-------
     2
(1 row)

  alter table pgmq.q_foo enable row level security;
  alter table pgmq.q_bar enable row level security;
  -- 0 issue - because RLS is active
  select * from lint."0019_insecure_queue_exposed_in_api";
 name | title | level | facing | categories | description | detail | remediation | metadata | cache_key 
------+-------+-------+--------+------------+-------------+--------+-------------+----------+-----------
(0 rows)

rollback;
